xonsh/history.py

```diff
@@ -0,0 +1,192 @@
+        while self.wait_for_shell:
+            time.sleep(0.01)
+        hsize, units = builtins.__xonsh_env__.get('XONSH_HISTORY_SIZE', (8128, 'files'))
+        files = self.unlocked_files()
+        # flag files for removal
+        if units == 'files':
+            rmfiles = files[:-hsize] if len(files) > hsize else []
+        elif units == 's':
+            now = time.time()
+            rmfiles = []
+            for ts, f in files:
+                if (now - ts) < hsize:
+                    break
+                rmfiles.append((None, f))
+        elif units == 'b':
+            n = 0
+            nbytes = 0
+            for f in files[::-1]:
+                fsize = os.stat(f).st_size
+                if nbytes + fsize > hsize:
+                    break
+                nbytes += fsize
+                n += 1
+            rmfiles = f[-n:]    
+        else:
+            raise ValueError('Units of {0!r} not understood'.format(unit))
+        # finally, clean up files
+        for _, f in rmfiles:
+            try:
+                os.remove(f)
+            except OSError:
+                pass
+
+    def unlocked_files(self):
+        """Finds the history files and returns the ones that are unlocked, this is 
+        sorted by the last closed time. Returns a list of (timestamp, file) tuples.
+        """
+        xdd = os.path.abspath(builtins.__xonsh_env__['XONSH_DATA_DIR'])
+        fs = [f for f in iglob(os.path.join(xdd, 'xonsh-*.json'))]
+        files = []
+        for f in fs:
+            try:
+                lj = lazyjson.LazyJSON(f, reopen=False)
+                if lj['locked']:
+                    continue
+                files.append((lj['ts'][1], f))
+                lj.close()
+            except (IOError, OSError):
+                continue
+        files.sort()
+        return files
+
+
+class HistoryFlusher(Thread):
+
+    def __init__(self, filename, buffer, queue, cond, at_exit=False, *args, **kwargs):
+        """Thread for flushing history."""
+        super(HistoryFlusher, self).__init__(*args, **kwargs)
+        self.filename = filename
+        self.buffer = buffer
+        self.queue = queue
+        queue.append(self)
+        self.cond = cond
+        self.at_exit = at_exit
+        if at_exit:
+            self.dump()
+        else:
+            self.start()
+
+    def run(self):
+        with self.cond:
+            self.cond.wait_for(self.i_am_at_the_front)
+            self.dump()
+            self.queue.popleft()
+
+    def i_am_at_the_front(self):
+        return self is self.queue[0]
+
+    def dump(self):
+        with open(self.filename, 'r') as f:
+            hist = lazyjson.LazyJSON(f).load()
+        hist['cmds'].extend(self.buffer)
+        if self.at_exit:
+            hist['ts'][1] = time.time()  # apply end time
+            hist['locked'] = False
+        with open(self.filename, 'w') as f:
+            lazyjson.dump(hist, f, sort_keys=True)
+
+
+class History(object):
```

Comment:
Maybe I am just being dense, but is there a clean way to get access to the elements in the history?  From what I could tell, I had to do:
```
from xonsh.lazyjson import LazyJSON
LazyJSON(__xonsh_history__.filename).load
```
to get something that looked like a Python object I could mess with.

Can we add indexing, etc to the ``History`` class directly to make accessing the history easier?